# Preparación Pre DB

Vamos a preparar los DataFrame para insertarlos en la Base de datos

In [3]:
# Importamos las librerías que necesitamos

# Librerías para automatización de navegadores (Selenium)
# -----------------------------------------------------------------------
from selenium import webdriver                   # Control automático del navegador para realizar scraping web

from selenium.webdriver.chrome.service import Service  # Maneja el servicio de ChromeDriver
from selenium.webdriver.common.by import By      # Para seleccionar elementos en la página web (DOM)
from selenium.webdriver.chrome.options import Options  # Configuración de opciones del navegador (headless mode, etc.)
import time                                      # Para gestionar pausas y temporización

# Librerías para extracción de datos y scraping
# -----------------------------------------------------------------------
from bs4 import BeautifulSoup  # Para analizar y extraer datos de HTML y XML (scraping web)

# Librerías para tratamiento y manipulación de datos
# -----------------------------------------------------------------------
import pandas as pd             # Para manipulación de estructuras de datos como DataFrames
import numpy as np              # Para cálculos numéricos y manejo de arrays
import datetime                 # Para manipulación de fechas y tiempos
# -----------------------------------------------------------------------
# Librerías para visualización de progreso
# -----------------------------------------------------------------------
from tqdm import tqdm  # Para crear barras de progreso durante los bucles

# Manejo de tiempo y solicitudes con espera
# -----------------------------------------------------------------------
from time import sleep          # Para pausar la ejecución del código por un tiempo determinado
import random                   # Para generar valores aleatorios, útil para espaciar solicitudes y evitar bloqueos
# Importación de módulos desde rutas recursivas
# -----------------------------------------------------------------------
import sys
sys.path.append("../")  # Añadir rutas externas para poder importar módulos desde carpetas superiores
from src.SupportPreDB import obtener_indice_supermercado

# Abrir tabla
Para continuar donde lo dejamos en la Fase 1

In [4]:
df_historicos =  pd.read_csv("../datos/01_tabla_historicos_preDB.csv",index_col="Unnamed: 0")

In [5]:
df_historicos

supermercado          categoria  \
0         mercadona  aceite de girasol   
1         mercadona  aceite de girasol   
2         mercadona  aceite de girasol   
3         mercadona  aceite de girasol   
4         mercadona  aceite de girasol   
...             ...                ...   
128791      alcampo  leche sin lactosa   
128792      alcampo  leche sin lactosa   
128793      alcampo  leche sin lactosa   
128794      alcampo  leche sin lactosa   
128795      alcampo  leche sin lactosa   

                                                 producto       fecha  \
0          Aceite De Girasol Refinado 0,2º Hacendado 1 L.  2024-07-12   
1          Aceite De Girasol Refinado 0,2º Hacendado 1 L.  2024-07-13   
2          Aceite De Girasol Refinado 0,2º Hacendado 1 L.  2024-07-14   
3          Aceite De Girasol Refinado 0,2º Hacendado 1 L.  2024-07-15   
4          Aceite De Girasol Refinado 0,2º Hacendado 1 L.  2024-07-16   
...                                                   ...         ...   
128791  Puleva Proteína Extra Pro Leche De Vaca Desnat...  2024-10-22   
128792  Puleva Proteína Extra Pro Leche De Vaca Desnat...  2024-10-23   
128793  Puleva Proteína Extra Pro Leche De Vaca Desnat...  2024-10-24   
128794  Puleva Proteína Extra Pro Leche De Vaca Desnat...  2024-10-25   
128795  Puleva Proteína Extra Pro Leche De Vaca Desnat...  2024-10-26   

        precio (€)  variacion (€)  variacion (%)  
0             1.45           0.00           0.00  
1             1.45           0.00           0.00  
2             1.45           0.00           0.00  
3             1.45           0.00           0.00  
4             1.45           0.00           0.00  
...            ...            ...            ...  
128791        7.74           0.00           0.00  
128792        7.74           0.00           0.00  
128793        7.74           0.00           0.00  
128794        7.74           0.00           0.00  
128795        9.00           1.26          16.28  

[128796 rows x 7 columns]

Ver si merece la pena tener una tabla producto

Con 1523 productos, nos merece la pena para no repetirlos en los 128796

In [6]:
len(df_historicos["producto"].unique())

1523

### Separar las tablas para realizar la base de Datos
Tendremos 4 tablas:
- Supermercado
- Categoría
- Producto
- Histórico

En la fase 3 detallaremos sus columnas y demás, por ahora solo queremos preparar los DataFrames

Vamos a realizar lo siguiente:
- Separar Supermercado, Categoría y Producto, cada uno en un DataFrame individual, con dos columnas, un index que empieza en uno, que compatibiliza con el funcionamiento de los índices en SQL y otra con los valores únicos de cada una de las columnas.

- DF histórico con índices de Supermercado, categoría y Producto: Generamos un diccionario para cada una de las tablas anteriores, con ello lo utilizaremos para reemplazar los valores de esas columnas por el índice que tendran en la Base de Datos, lo que nos ayudará a vincular las Foreign Keys